src/libguac/user-handshake.c

```diff
@@ -305,96 +379,53 @@ int guac_user_handle_connection(guac_user* user, int usec_timeout) {
-    if (guac_parser_expect(parser, socket, usec_timeout, "size")) {
-
-        /* Log error */
-        guac_user_log_handshake_failure(user);
-        guac_user_log_guac_error(user, GUAC_LOG_DEBUG,
-                "Error reading \"size\"");
-
-        guac_parser_free(parser);
-        return 1;
-    }
-
-    /* Validate content of size instruction */
-    if (parser->argc < 2) {
-        guac_user_log(user, GUAC_LOG_ERROR, "Received \"size\" "
-                "instruction lacked required arguments.");
-        guac_parser_free(parser);
-        return 1;
-    }
-
-    /* Parse optimal screen dimensions from size instruction */
-    user->info.optimal_width  = atoi(parser->argv[0]);
-    user->info.optimal_height = atoi(parser->argv[1]);
-
-    /* If DPI given, set the user resolution */
-    if (parser->argc >= 3)
-        user->info.optimal_resolution = atoi(parser->argv[2]);
-
-    /* Otherwise, use a safe default for rough backwards compatibility */
-    else
-        user->info.optimal_resolution = 96;
-
-    /* Get supported audio formats */
-    if (guac_parser_expect(parser, socket, usec_timeout, "audio")) {
-
-        /* Log error */
-        guac_user_log_handshake_failure(user);
-        guac_user_log_guac_error(user, GUAC_LOG_DEBUG,
-                "Error reading \"audio\"");
-
-        guac_parser_free(parser);
-        return 1;
-    }
-
-    /* Store audio mimetypes */
-    char** audio_mimetypes = guac_copy_mimetypes(parser->argv, parser->argc);
-    user->info.audio_mimetypes = (const char**) audio_mimetypes;
-
-    /* Get supported video formats */
-    if (guac_parser_expect(parser, socket, usec_timeout, "video")) {
-
-        /* Log error */
-        guac_user_log_handshake_failure(user);
-        guac_user_log_guac_error(user, GUAC_LOG_DEBUG,
-                "Error reading \"video\"");
-
-        guac_parser_free(parser);
-        return 1;
-    }
-
-    /* Store video mimetypes */
-    char** video_mimetypes = guac_copy_mimetypes(parser->argv, parser->argc);
-    user->info.video_mimetypes = (const char**) video_mimetypes;
-
-    /* Get supported image formats */
-    if (guac_parser_expect(parser, socket, usec_timeout, "image")) {
-
-        /* Log error */
-        guac_user_log_handshake_failure(user);
-        guac_user_log_guac_error(user, GUAC_LOG_DEBUG,
-                "Error reading \"image\"");
-
-        guac_parser_free(parser);
-        return 1;
-    }
-
-    /* Store image mimetypes */
-    char** image_mimetypes = guac_copy_mimetypes(parser->argv, parser->argc);
-    user->info.image_mimetypes = (const char**) image_mimetypes;
-
-    /* Get args from connect instruction */
-    if (guac_parser_expect(parser, socket, usec_timeout, "connect")) {
-
-        /* Log error */
-        guac_user_log_handshake_failure(user);
-        guac_user_log_guac_error(user, GUAC_LOG_DEBUG,
-                "Error reading \"connect\"");
+    /* Handle each of the opcodes. */
+    while (1) {
```

Comment:
Conditionless `while` loops make me itchy ... but I imagine this may vanish regardless if the duplicate code between the main user instruction handlers and the handshake user instruction handlers is separated out. If unavoidable, then unavoidable, but it would be better to have an explicit bound on the loop.

Reply:
Should be able to consolidate stuff.  I actually had it done, but was running into issues and realized I needed to change fewer things concurrently.  So, now I'll go back and do this.  I was having some issues with includes, so I might end up needing some help - it's a little challenging with the maps to get everything in the correct header files to be seen where it needs to be.

Reply:
Well, didn't manage to make the conditionless while loop disappear, yet, so I'll be happy to take a suggestion or two on the best way to do this.

Reply:
I think the usual structure would be (pseudocode):

    while (instruction successfully read) {
        handle_success;
    }

    handle_failure;

In our case, though, we have more like:

    while (handshake_in_progress) {
        handle_handshake_instruction;
    }

    this_might_be_failure_or_success;

I'm not sure myself how this could be arranged as-is to remove the unbounded loop, given that the exit condition of the loop would be ambiguously either success or failure. Things might be easier/cleaner if the actual internal handshake handling were its own function. Then you'd end up with something like:

    if (init_user_with_handshake(stuff))
        handle_failure;

    add_user_to_connection

with the actual function being the expected:

    while (instruction successfully read) {

        if (instruction_is_connect)
            return 0;

        handle_instruction;
    }

    handle_failure;
    return 1;

I'm sure there's a way to rearrange the logic such that we have a net gain in readability, part of that readability being an explicit loop bound.

Reply:
Okay, I'll go another round with that one and see what I can do.

Reply:
Okay, this should be implemented, now, in a way that gets rid of the infinite loop, breaking the handshake out into a separate function.